File: fsu/monte-carlo-methods/goodness-of-fit/code/fibonacci.hs
main = print (kolmogorovSmirnov fibonacci 1000)

fibonacci :: [Int]
fibonacci = 1:1:[n | i <- [2..], let n = mod (fibonacci !! (i-1) + fibonacci !! (i-2)) (2 ^ 31)]

kolmogorovSmirnov seq n = max (maximum dPlus) (maximum dMinus)
    where dPlus = [(fromIntegral k) / (fromIntegral n) - fromIntegral (seq' !! (k - 1)) | k <- [1..n]]
          dMinus = [fromIntegral (seq' !! (k - 1)) - (fromIntegral (k - 1)) / (fromIntegral n) | k <- [1..n]]
          seq' = take n seq

File: fsu/monte-carlo-methods/goodness-of-fit/code/fibonacci.py
import numpy as np

def fibonacci(n):
    f = [1, 1]
    for i in range(0, n-2):
        f.append((f[i] + f[i+1]) % pow(2, 31))
    return f

def kolmogorov_smirnov(seq):
    n = len(seq)
    Dplus = np.max([(k+1) / len(seq) - seq[k] for k in range(0, len(seq))])
    Dminus = np.max([seq[k] - k / len(seq) for k in range(0, len(seq))])
    return max(Dplus, Dminus)

def main():
    print(kolmogorov_smirnov(fibonacci(1000)))

main()
